src/shell/shell.cpp

```diff
@@ -42,7 +42,7 @@ callback_number_t call_shellstop = 0;
 /* Larger scope so shell_del autoexec can use it to
  * remove things from the environment */
 DOS_Shell *first_shell = nullptr;
-std::unique_ptr<ShellHistory> global_shell_history;
+ShellHistory global_shell_history = {};
```

Comment:
Okay, so no globals like this please going forward. We want to remove this stuff, not add more 😄 

We should not rely on the init sequence of globals and static vars. Having global vars are fine as long as they're initialised manually by the code somewhere, so the init order is clear.

Reply:
Originally the history was a member variable of the `Shell` class but @MeAreJeenius moved it to this global in his rework.  I believe the reasoning was so that we could save the history of child shells as well.  My original implementation only saved the history from `first_shell`.

In this case, init order shouldn't be a problem.  I've removed the constructors and destructors so no worries there.  It's also always in a valid state.  The data members are just a `std::vector<std::string>` and a `std_fs::path`.  When both of those are default constructed, it just means an empty history and no history file (same as if the user set `shell_history_file` to empty in their config file).

I think this is better than the smart pointer.  Before, it was possible (although probably very unlikely in practice) for a dangling reference to be held by some `Shell` object or to have a nullptr de-reference if a `Shell` happened to be created before `global_shell_history`.

IMO if we need a global like this, make a real global but make it as Plain Old Data as possible so we don't run into the issues you mentioned.

Reply:
@weirddan455 I get your point, it's just Plain Old Data approach is error prone as you need to be "careful". I mean, for every single global you need to check whether it's safe or not, while if you nuke all globals, then the issue is non-existant.

Similarly to always using copies, always defaulting to immutable structures and mutate only when necessary etc., then things are just thread-safe by default, etc.

But ok, if it's too difficult in this case to remove the global, it's fine.

Reply:
I get that globals should be used sparingly although nuking all globals from the codebase would be challenging to say the least.

In this case, I just really can't think of a better approach.  The current code uses a `std::unique_ptr` but then bypasses the safety of that by de-referencing the smart pointer and storing an un-safe reference as a member variable.  The global is certainly safer than that.

A `std::shared_ptr` would be a much better thing to use but then it's tricky of when to construct the singleton object.  You could do something like this:

```cpp
static std::shared_ptr<ShellHistory> maybe_create_history()
{
    static std::shared_ptr<ShellHistory> global_history = {};
    if (!global_history)
    {
        global_history = std::make_shared();
    }
    return global_history;
}

class Shell
{
    std::shared_ptr<ShellHistory> global_history = {};

    Shell::Shell()
    {
        global_history = maybe_create_history();
    }
}
```

But then that's just a global with extra steps really.

If you have a better idea, feel free to share but I honestly can't think of one.

Reply:
> In this case, I just really can't think of a better approach. The current code uses a `std::unique_ptr` but then bypasses the safety of that by de-referencing the smart pointer and storing an un-safe reference as a member variable. The global is certainly safer than that.

Yeah, okay, let's not overdo it 😄 We should not choose solutions based on ideology, agreed, rather than to solve the problem at hand with minimal fuss.

I'll look into weak pointers at some points as I have no idea what that is 😄

Reply:
Weak pointer is a non-owning reference to a shared pointer.  I just put in a 2nd PR using weak pointer and keeping RAII and you can let me know which one you like better.  See #3664